### PR TITLE
Stack Trace Decoration for Beam Samza Runner

### PR DESCRIPTION
**Pull request by Sanil15**
Feature:

1. Added support for displaying the full tranform name with any stack-trace for Samza Runner
2. Added support for decorating stack traces with additional debugging information by adding a SamzaExceptionListener via the observer pattern

Example: Full transform name will be displayed with stack-trace on any exception thrown at runtime for ex
```

2022-08-02 11:45:30 ERROR [OpAdapter] [Samza StreamProcessor Container Thread-0] Op My Custom Operator/ParMultiDo(Anonymous) threw an exception during processing
java.lang.RuntimeException: org.apache.beam.sdk.util.UserCodeException: java.lang.RuntimeException: My Custom Exception
	at org.apache.beam.runners.samza.metrics.DoFnRunnerWithMetrics.withMetrics(DoFnRunnerWithMetrics.java:96) ~[beam-runners-samza-200.30.0.5.jar:?]
	at org.apache.beam.runners.samza.metrics.DoFnRunnerWithMetrics.processElement(DoFnRunnerWithMetrics.java:53) ~[beam-runners-samza-200.30.0.5.jar:?]
	at org.apache.beam.runners.core.SimplePushbackSideInputDoFnRunner.processElementInReadyWindows(SimplePushbackSideInputDoFnRunner.java:79) ~[beam-runners-core-java-200.30.0.5.jar:?]
	at org.apache.beam.runners.samza.runtime.DoFnOp.processElement(DoFnOp.java:295) ~[beam-runners-samza-200.30.0.5.jar:?]
	at org.apache.beam.runners.samza.runtime.OpAdapter.apply(OpAdapter.java:94) ~[beam-runners-samza-200.30.0.5.jar:?]


```
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/OpAdapter.java [hunk 1]
```diff
@@ -102,7 +115,16 @@ public synchronized CompletionStage<Collection<OpMessage<OutT>>> apply(OpMessage
               String.format("Unexpected input type: %s", message.getType()));
       }
     } catch (Exception e) {
-      LOG.error("Op {} threw an exception during processing", this.getClass().getName(), e);
+      LOG.error("Op {} threw an exception during processing", transformFullName, e);
+      e.addSuppressed(
+          new RuntimeException(
+              String.format("Op: %s threw an exception during processing", transformFullName)));
+      exceptionListeners.forEach(
```
Comment (xinyuiscool) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/OpAdapter.java:122]: Please add another try-catch {ignore} here for exceptionListeners.forEach. I don't want this block to be interrupted by the listeners.

You might consider move this logic line 119 - 127 in a private function, e.g. notifyExceptionListners(...)

**Review by xinyuiscool — COMMENTED** [review]
Overall looks good. A minor readability improvement suggestion below.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/OpAdapter.java [hunk 2]
```diff
@@ -102,7 +115,16 @@ public synchronized CompletionStage<Collection<OpMessage<OutT>>> apply(OpMessage
               String.format("Unexpected input type: %s", message.getType()));
       }
     } catch (Exception e) {
-      LOG.error("Op {} threw an exception during processing", this.getClass().getName(), e);
+      LOG.error("Op {} threw an exception during processing", transformFullName, e);
+      e.addSuppressed(
+          new RuntimeException(
+              String.format("Op: %s threw an exception during processing", transformFullName)));
```
Comment (xinyuiscool) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/OpAdapter.java:121]: Wording: "Exception happened in transform %s", ...

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/OpAdapter.java (hunk 1 above)
Comment (Sanil15) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/OpAdapter.java]: good call, done

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/OpAdapter.java (hunk 2 above)
Comment (Sanil15) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/OpAdapter.java]: done

**Review by xinyuiscool — APPROVED** [review]
LGTM!

**Comment by xinyuiscool** [comment]
Run Java PreCommit

**Comment by xinyuiscool** [comment]
Run Java PreCommit

**Pull request merged**